### PR TITLE
Changes for TrafficManager 2018-04-01 API

### PR DESCRIPTION
**Pull request by allencal**
Powershell changes for TrafficManager 2018-04-01 API

## Description

PowerShell changes for TrafficManager 2018-04-01 API

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by markcowl** [comment]
@allencal Can you provide a link to your cmdlet review?

**Comment by allencal** [comment]
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/111

**Comment by markcowl** [comment]
@allencal You need to publish your SDK package - you should not use local packages for changes in the preview branch.

**Comment by praries880** [comment]
@allencal Marking this as Do Not Merge, well remove it once the SDK is released.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/Add-AzureRmTrafficManagerEndpointConfig.md [hunk 1]
```diff
@@ -39,13 +42,28 @@ PS C:\> Set-AzureRmTrafficManagerProfile -TrafficManagerProfile $TrafficManagerP
...
 The second command adds an endpoint named contoso to the profile stored in $TrafficManagerProfile.
 The command includes configuration data for the endpoint.
 This command changes only the local object.
+
 The final command updates the Traffic Manager profile in Azure to match the local value in $TrafficManagerProfile.
 
 ## PARAMETERS
 
+### -CustomHeaders
+List of custom header name and value pairs for probe requests.```yaml
```
Comment (markcowl) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/Add-AzureRmTrafficManagerEndpointConfig.md]: This is one of the trailing yaml string mentioned in the validatehelpissues report.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/New-AzureRmTrafficManagerProfile.md [hunk 2]
```diff
@@ -42,6 +46,19 @@ The DNS FQDN is contosoapp.trafficmanager.net.
 
 ## PARAMETERS
 
+### -CustomHeaders
+List of custom header name and value pairs for probe requests.```yaml
```
Comment (markcowl) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/New-AzureRmTrafficManagerProfile.md]: This is one of the trailing yaml strings mentioned in the validatehelpissues report

Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/New-AzureRmTrafficManagerProfile.md]: Yes, I see them in 3 files. Will fix.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ExpectedStatusCodeRange/RemoveAzureTrafficManagerExpectedStatusCodeRange.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+namespace Microsoft.Azure.Commands.TrafficManager
+{
+    using System;
+
+    [Cmdlet(VerbsCommon.Remove, "AzureRmTrafficManagerExpectedStatusCodeRange"), OutputType(typeof(TrafficManagerProfile))]
+    public class RemoveAzureTrafficManagerExpectedStatusCodeRange : TrafficManagerBaseCmdlet
+    {
+        [Parameter(Mandatory = true, HelpMessage = "The minimum value of the expected status code range to be removed.", ParameterSetName = "Fields")]
+        [ValidateNotNullOrEmpty]
+        public int? Min { get; set; }
```
Comment (praries880) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/ExpectedStatusCodeRange/RemoveAzureTrafficManagerExpectedStatusCodeRange.cs]: Since the param is mandatory, it PS will make sure a  value is specified for the param, it can not be null.

Instead of marking the param as a nullable int, id leave it as an int and instead of [ValidateNotNullOrEmpty] maybe use [ValidateRange] is there is a certain range the values will be in.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1 [hunk 4]
```diff
@@ -654,4 +654,99 @@ function Test-PipeEndpointFromGetProfile
...
+	
+    $retrievedEndpoint = $retrievedProfile.Endpoints[0]
+
+	Assert-True { Add-AzureRmTrafficManagerCustomHeaderToEndpoint -Name "foo" -Value "bar" -TrafficManagerEndpoint $retrievedEndpoint }
+
+    Set-AzureRmTrafficManagerEndpoint -TrafficManagerEndpoint $retrievedEndpoint
+
+    $endpoint = Get-AzureRmTrafficManagerEndpoint -Name $endpointName -ProfileName $profileName -ResourceGroupName $resourceGroup.ResourceGroupName -Type "ExternalEndpoints"
+
+	Assert-AreEqual "foo" $endpoint.CustomHeaders[0].Name
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1]: Multiple custom headers are allowed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/Add-AzureRmTrafficManagerEndpointConfig.md (hunk 1 above)
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/Add-AzureRmTrafficManagerEndpointConfig.md]: Fixed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/New-AzureRmTrafficManagerProfile.md (hunk 2 above)
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/help/New-AzureRmTrafficManagerProfile.md]: Fixed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToEndpoint.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        [Parameter(Mandatory = true, HelpMessage = "The name of the custom header.", ParameterSetName = "Fields")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "The value of the custom header.", ParameterSetName = "Fields")]
+        [ValidateNotNullOrEmpty]
+        public string Value { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, HelpMessage = "The endpoint the header will be added to.")]
+        [ValidateNotNullOrEmpty]
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToEndpoint.cs]: Why? Should an empty custom header value be permitted?

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToProfile.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        [Parameter(Mandatory = true, HelpMessage = "The name of the custom header.", ParameterSetName = "Fields")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "The value of the custom header.", ParameterSetName = "Fields")]
+        [ValidateNotNullOrEmpty]
+        public string Value { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, HelpMessage = "The profile the header will be added to.")]
+        [ValidateNotNullOrEmpty]
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToProfile.cs]: Seems to me it should remain ValidateNotNullOrEmpty.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md [hunk 7]
```diff
@@ -18,9 +18,16 @@
         - Additional information about change #1
 -->
 ## Current Release
+* Support for new routing methods and features
 
 ## Version 4.0.8
-* Updated help files to include full parameter types and correct input/output types.
+* Support for the MultiValue routing method
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md]: Fixed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/AzureRM.TrafficManager.Netcore.psd1 [hunk 8]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '0.13.1'
+ModuleVersion = '0.14.0'
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/AzureRM.TrafficManager.Netcore.psd1]: Done

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/AzureRM.TrafficManager.Netcore.psd1 [hunk 9]
```diff
@@ -51,7 +51,7 @@ PowerShellVersion = '5.1'
 # ProcessorArchitecture = ''
 
 # Modules that must be imported into the global environment prior to importing this module
-RequiredModules = @(@{ModuleName = 'AzureRM.Profile.Netcore'; ModuleVersion = '0.13.1'; })
+RequiredModules = @(@{ModuleName = 'AzureRM.Profile.Netcore'; ModuleVersion = '0.14.0'; })
```
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/AzureRM.TrafficManager.Netcore.psd1]: Done

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1]: I see what you mean... Not if it means re-generating the session records. We have Assert-AreEqual 0 $endpoint.CustomHeaders.Count after deleting it, so it's implicitly covered already.

**Comment by praries880** [comment]
@allencal There are a bunch of failed tests : https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/6410/artifact/src/Publish/TestResults/FailingTests/Microsoft.Azure.Commands.TrafficManager.Test.ScenarioTests.dll.html

Can you look at them and fix them.

**Comment by allencal** [comment]
Failing tests fixed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2.Test/ScenarioTests/EndpointTests.ps1]: Done.

**Comment by praries880** [comment]
@allencal 
Did you hand author the help files or did you follow the instructions [here](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)
If not, i'd suggest deleting the hand authored changes and regenerating the help files using the steps mentioned in the guide.

**Comment by allencal** [comment]
Tests updated.

File: documentation/development-docs/azure-powershell-developer-guide.md [hunk 10]
```diff
@@ -57,6 +57,12 @@ The following prerequisites should be completed before contributing to the Azure
     - `C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe`
     - `C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe`
     
+**Note** The PowerShell Build requires the 64-bit version of MSBuild that ships with visual studio, so ensure that this version is the first version in your cmd path.  64-bit MSBuild could be installed in a couple of places on your machine:
```
Comment (praries880) [inline documentation/development-docs/azure-powershell-developer-guide.md]: not sure how this change got in, but its already part of preview ,kindly undo this.

File: documentation/development-docs/help-generation.md [hunk 11]
```diff
@@ -72,6 +72,8 @@ Get-ChildItem -Path $PathToHelpFolder | Update-MarkdownHelp -AlphabeticParamsOrd
 
 This will update all of the markdown files with public interface changes made to corresponding cmdlets, add markdown files for any new cmdlets, remove markdown files for any deleted cmdlets, and update the module page (_e.g.,_ `AzureRM.Profile.md`) with any added or removed cmdlets.
 
+_This seems to work better when run from within the `help` folder itself (For e.g. to generate the help files for the [`Network`](https://github.com/Azure/azure-powershell/tree/preview/src/ResourceManager/Network) module run the cmd from under [`Commands.Network/help`](https://github.com/Azure/azure-powershell/tree/preview/src/ResourceManager/Network/Commands.Network/help)). Also, you will have to import the profile module from under <Repo base path>/src/Package/Debug/ResourceManager/AzureResourceManager/AzureRM.Profile/AzureRM.Profile.psd1_
```
Comment (praries880) [inline documentation/development-docs/help-generation.md]: Same here...

File: tools/CredScanSuppressions.json [hunk 12]
```diff
@@ -1,6 +1,10 @@
 {
     "tool": "Credential Scanner",
         "suppressions": [
+        {
+            "file": "\\documentation\\development-docs\\azure-powershell-developer-guide.md",
```
Comment (praries880) [inline tools/CredScanSuppressions.json]: why is this change needed?

**Comment by praries880** [comment]
on demand : https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/1038/

**Comment by allencal** [comment]
Merge conflicts fixed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/Az.TrafficManager.psd1 [hunk 13]
```diff
@@ -72,19 +72,28 @@ NestedModules = @('.\Microsoft.Azure.Commands.TrafficManager.dll')
...
-    'Remove-AzTrafficManagerEndpoint', 
-    'New-AzTrafficManagerEndpoint', 
-    'Remove-AzTrafficManagerEndpointConfig', 
-    'Add-AzTrafficManagerEndpointConfig', 
-    'Disable-AzTrafficManagerProfile', 
-    'Enable-AzTrafficManagerProfile', 
-    'Remove-AzTrafficManagerProfile', 
-    'Set-AzTrafficManagerProfile', 
-    'Get-AzTrafficManagerProfile', 'New-AzTrafficManagerProfile'
+CmdletsToExport = 'Add-AzureRmTrafficManagerCustomHeaderToEndpoint',
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/Az.TrafficManager.psd1]: You will need to have the Az syntax above

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToEndpoint.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+using System.Management.Automation;
+using ProjectResources = Microsoft.Azure.Commands.TrafficManager.Properties.Resources;
+
+namespace Microsoft.Azure.Commands.TrafficManager
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    [Cmdlet(VerbsCommon.Add, "AzureRmTrafficManagerCustomHeaderToEndpoint", SupportsShouldProcess = false), OutputType(typeof(TrafficManagerEndpoint))]
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/CustomHeader/AddAzureTrafficManagerCustomHeaderToEndpoint.cs]: All new cmdlets will need to have this cmdlet declaration:
```
[Cmdlet(VerbsCommon.Add, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "TrafficManagerCustomHeaderToEndpoint"
```
in order to support both Desktop and Netstandard.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md [hunk 15]
```diff
@@ -17,15 +17,18 @@
     * Overview of change #1
         - Additional information about change #1
 -->
-## Current Release
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md]: You need the ## Current release header

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md [hunk 16]
```diff
@@ -17,15 +17,18 @@
...
+    - Support for IP address ranges (subnets) in endpoints
+* Support for Custom Headers in profiles
+* Support for Expected status code ranges in profiles
+* Support for Custom Headers in endpoints
 * Fixed issue with default resource groups not being set.
 
 ## Version 4.0.9
 * Updated to the latest version of the Azure ClientRuntime.
 
-## Version 4.0.8
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md]: Why did you delete this?

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ExpectedStatusCodeRange/RemoveAzureTrafficManagerExpectedStatusCodeRange.cs [hunk 17]
```diff
@@ -0,0 +1,67 @@
+﻿//using System;
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/ExpectedStatusCodeRange/RemoveAzureTrafficManagerExpectedStatusCodeRange.cs]: Delete everything above the copyright message

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/packages.config [hunk 18]
```diff
@@ -1,4 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Microsoft.Azure.Management.TrafficManager" version="2.3.0-preview" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.TrafficManager" version="2.5.0" targetFramework="net452" />
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/packages.config]: You will need to remove the changes to tools/AzureRM/AzureRM.psm1 in the file below this one.

Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/packages.config]: Done.

**Comment by allencal** [comment]
New post-merge issues addressed.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md [hunk 19]
```diff
@@ -18,13 +18,17 @@
         - Additional information about change #1
 -->
 ## Current Release
-* Fixed issue with default resource groups not being set.
-
-## Version 4.0.9
 * Updated to the latest version of the Azure ClientRuntime.
 
 ## Version 4.0.8
-* Updated help files to include full parameter types and correct input/output types.
```
Comment (maddieclayton) [inline src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md]: I'm not sure what happened here, please revert all changes and just add your message under "Current Release"

File: tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv [hunk 20]
```diff
@@ -18,3 +18,11 @@
...
 "C:\azure-powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.TrafficManager\Microsoft.Azure.Commands.TrafficManager.dll","Microsoft.Azure.Commands.TrafficManager.NewAzureTrafficManagerProfile","New-AzureRmTrafficManagerProfile","1","8410","Parameter MonitorToleratedNumberOfFailures of cmdlet New-AzureRmTrafficManagerProfile does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
+"C:\azure-powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.TrafficManager\Microsoft.Azure.Commands.TrafficManager.dll","Microsoft.Azure.Commands.TrafficManager.AddAzureTrafficManagerCustomHeaderToEndpoint","Add-AzureRmTrafficManagerCustomHeaderToEndpoint","1","8100","Add-AzureRmTrafficManagerCustomHeaderToEndpoint Does not support ShouldProcess but the cmdlet verb Add indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
```
Comment (maddieclayton) [inline tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv]: All of these new cmdlets that make changes to the server should have supportshouldprocess = true.  This will allow users to pass in -Confirm which prompts them for confirmation upon run.  See here for example: https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/KeyVault/Commands.KeyVault/Commands/AddAzureKeyVaultCertificateContact.cs#L101

Comment (allencal) [reply tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv]: These are objects with two parameters that can be easily added and removed. Requiring verification would be a nuisance to the user and do more harm than good.

File: src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md (hunk 19 above)
Comment (allencal) [reply src/ResourceManager/TrafficManager/Commands.TrafficManager2/ChangeLog.md]: Making my best guess here; hope this is correct. I thought version 4.0.9 was the current release, didn't think both tags were supposed to be present.

File: tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv (hunk 20 above)
Comment (maddieclayton) [reply tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv]: The ShouldProcess loop only prompts the user if the user specifies -Confirm.  This gives the user flexibility to run cmdlets and see their possible effects before making any changes to the server.  This is a PowerShell requirement for all cmdlets that make modifications.

Comment (allencal) [reply tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv]: Done. From what I've read it will not have any effect unless the customer uses -Confirm, so it won't get in the way.

Comment (maddieclayton) [reply tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv]: @allencal Looks like you enabled supportsshouldprocess correctly, but you also need to add a loop around the modification in each of the cmdlets.  See this for example of loop (line 101): https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/KeyVault/Commands.KeyVault/Commands/AddAzureKeyVaultCertificateContact.cs#L101.  Once that is done, you can remove all the these additional lines in tools/StaticAnalysis/Exceptions/AzureRM.TrafficManager/SignatureIssues.csv.

**Comment by allencal** [comment]
Support for ShouldProcess added.
